src/backchannel_logout.rs

```diff
@@ -0,0 +1,469 @@
+    Audience, IssuerUrl, SubjectIdentifier,
+};
+
+/// The Logout Token as defined in [section 2.4] of the [OpenID Connect Back-Channel Logout spec][1]
+///
+/// [section 2.4]: <https://openid.net/specs/openid-connect-backchannel-1_0.html#LogoutToken>
+/// [1]: <https://openid.net/specs/openid-connect-backchannel-1_0.html>
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct LogoutToken {
+    /// The issuer of this token
+    iss: IssuerUrl,
+    /// The audience this token is intended for
+    aud: Vec<Audience>,
+    /// Time at which this token was issued
+    iat: DateTime<Utc>,
+    /// The unique identifier for this token. This can be used to detect
+    /// replay attacks.
+    jti: String,
+    identifier: Identifier,
+    events: HashMap<String, serde_json::Value>,
+}
+
+impl LogoutToken {
+    /// The `iss` claim
+    pub fn issuer(&self) -> &IssuerUrl {
+        &self.iss
+    }
+
+    /// The `aud` claim
+    pub fn audiences(&self) -> impl Iterator<Item = &Audience> {
+        self.aud.iter()
+    }
+
+    /// The `iat` claim
+    pub fn issue_time(&self) -> DateTime<Utc> {
+        self.iat
+    }
+
+    /// The `jti` claim. It's the unique identifier for this token and can be
+    /// used to detect replay attacks.
+    pub fn jti(&self) -> &str {
+        &self.jti
+    }
+
+    /// As per spec, a [`LogoutToken`] MUST either have the `sub`  or `sid`
+    /// claim and MAY contain both. You can match the [`Identifier`] to detect
+    /// which claims are present.
+    pub fn identifier(&self) -> &Identifier {
+        &self.identifier
+    }
+
+    /// A [`LogoutToken`] is compatible with the [SET standard from RFC 8417][1]
+    ///
+    /// [1]: <https://www.rfc-editor.org/info/rfc8417>
+    pub fn events(&self) -> &HashMap<String, serde_json::Value> {
+        &self.events
+    }
+}
+impl<'de> Deserialize<'de> for LogoutToken {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        let value: serde_json::Value = serde_json::Value::deserialize(deserializer)?;
+        if let serde_json::Value::Object(ref map) = value {
+            if map.contains_key("nonce") {
+                return Err(<D::Error as Error>::custom("nonce claim is prohibited"));
+            }
+        }
+
+        #[derive(Deserialize)]
+        struct Repr {
+            /// The issuer of this token
+            iss: IssuerUrl,
+            /// The audience this token is intended for
+            #[serde(deserialize_with = "deserialize_string_or_vec")]
+            aud: Vec<Audience>,
+            /// Time at which this token was issued
+            #[serde(with = "serde_utc_seconds")]
+            iat: DateTime<Utc>,
+            /// The unique identifier for this token. This can be used to detect
+            /// replay attacks.
+            jti: String,
+            #[serde(flatten)]
+            identifier: Identifier,
+            events: HashMap<String, serde_json::Value>,
```

Comment:
Using types, we can have serde check the URL key for us:
```rust

#[derive(Clone, Debug, Deserialize, Serialize)]
#[serde(deny_unknown_fields)]
struct BackchannelLogoutEvent {}

#[derive(Clone, Debug, Deserialize, Serialize)]
struct LogoutTokenEvents {
  #[serde(rename = "http://schemas.openid.net/event/backchannel-logout")]
  backchannel_logout: BackchannelLogoutEvent,
}
```

Reply:
Yes, I thought about this and came to the conclusion that it's not a good idea if not done properly. But in order to to this right, I guess we need to completely implement the [SET](https://www.rfc-editor.org/info/rfc8417) standard and have the different events typed up in something like an enum. Also, I think this

```rust
#[derive(Clone, Debug, Deserialize, Serialize)]
#[serde(deny_unknown_fields)] // <--- would violate the spec
struct BackchannelLogoutEvent {}
```

since the spec says about this ([Section 2.4](https://openid.net/specs/openid-connect-backchannel-1_0.html#LogoutToken)): 
> The corresponding member value MUST be a JSON object and **SHOULD** be the empty JSON object {}. 

it only SHOULD not MUST be empty, so we can't deny unknown fields, can we?

I've only added the check [here](https://github.com/ramosbugs/openidconnect-rs/pull/70/commits/091b675b0322154b8d5c0ddf0d3324589bc2c344#diff-94f740fa2f6aa14eea6873b596009047af7f9caef16d6ab80a619358233a7c7cR106) to ensure that we only parse tokens that are valid to the spec. I've even considered completely ignoring this field apart from making sure it complies to the spec.

Reply:
Oops, you are correct! Let's remove `deny_unknown_fields` then. What if we leave the `BackchannelLogoutEvent` struct empty for now but use `#[non_exhaustive]` so that we can add fields in the future without a breaking change?

Reply:
Wouldn't it make sense to use a enum for `LogoutTokenEvents`?

Reply:
Enums seem appropriate when we need to distinguish between mutually-exclusive types. In this case, don't we know that it should always contain an event with the `http://schemas.openid.net/event/backchannel-logout` key? What other info do you envision needing to represent with `LogoutTokenEvents`?

If extensibility is the issue, we could add a generic flattened `additional_claims` field. Since there aren't any other flattened fields here, we won't need `FilteredFlatten`.

Similarly, we could make `backchannel_logout` a generic type instead of defining it as the empty `BackchannelLogoutEvent` struct. It depends how extensible we want to get and what future functionality users might want without needing to introduce breaking changes.

Reply:
The thing with this is that it might include other "events" than `http://schemas.openid.net/event/backchannel-logout` 
and as much as I'd like to have all of these typed in an enum or something, it's out of scope for this crate.

I think just exposing the HashMap like it is currently should be fine since I don't think anyone would actually care about this and even if someone does, they could just use `serde_json::from_value` to parse their structs.

> What other info do you envision needing to represent with LogoutTokenEvents?

Forget that, I thought we could do something like this, but that ain't gonna work:

```rust
enum Event {
    #[serde(rename = "http://schemas.openid.net/event/backchannel-logout")]
    Backchannel {},
    // other possible SET event
    #[serde(rename = "https://schemas.openid.net/secevent/risc/event-type/account-disabled")]
    AccountDisabled {
         reason: String
    }
}
```